contracts/random/RandomValues.sol

```diff
@@ -0,0 +1,127 @@
+
+/**
+ * @notice Game contracts that need random numbers should extend this contract.
+ * @dev This contract can be used with UPGRADEABLE or NON-UNGRADEABLE contracts.
+ */
+abstract contract RandomValues {
+    // Address of random seed provider contract.
+    // This value "immutable", and hence patched directly into bytecode when it is used.
+    // There will only ever be one random seed provider per chain. Hence, this value
+    // does not need to be changed.
+    RandomSeedProvider public immutable randomSeedProvider;
+
+    // Map of request id to fulfilment id.
+    mapping(uint256 requestId => uint256 fulfilmentId) private randCreationRequests;
+    // Map of request id to random source. Retaining the source allows for upgrade
+    // of sources inside the random seed provider contract.
+    mapping(uint256 requestId => address randomSource) private randCreationRequestsSource;
+
+    // Each request has a unique request id. The id is the current value
+    // of nextNonce. nextNonce is incremented for each request.
+    uint256 private nextNonce;
+
+    /**
+     * @notice Set the address of the random seed provider.
+     * @param _randomSeedProvider Address of random seed provider.
+     */
+    constructor(address _randomSeedProvider) {
+        randomSeedProvider = RandomSeedProvider(_randomSeedProvider);
+    }
+
+    /**
+     * @notice Register a request to generate a random value. This function should be called
+     *      when a game player has purchased an item the value of which is based on a random value.
+     * @return _randomRequestId A value that needs to be presented when fetching the random
+     *      value with fetchRandom.
+     */
+    function _requestRandomValueCreation() internal returns (uint256 _randomRequestId) {
+        uint256 randomFulfillmentIndex;
+        address randomSource;
+        (randomFulfillmentIndex, randomSource) = randomSeedProvider.requestRandomSeed();
+        _randomRequestId = nextNonce++;
+        randCreationRequests[_randomRequestId] = randomFulfillmentIndex;
+        randCreationRequestsSource[_randomRequestId] = randomSource;
+    }
+
+    /**
+     * @notice Fetch a random value that was requested using _requestRandomValueCreation.
+     * @dev The value is customised to this game, the game player, and the request by the game player.
+     *      This level of personalisation ensures that no two players end up with the same random value
+     *      and no game player will have the same random value twice.
+     * @param _randomRequestId The value returned by _requestRandomValueCreation.
+     * @return _randomValue The random number that the game can use.
+     */
+    function _fetchRandom(uint256 _randomRequestId) internal returns (bytes32 _randomValue) {
+        // Don't return the personalised seed directly. Otherwise there is a risk that
+        // the seed will be revealed, which would then compromise the security of calls
```

Comment:
Can you elaborate on the issue here? The seed could be found in the memory trace from trace apis. Is the idea that It can't be revealed within this transaction, but afterwards is fine?

Reply:
If _fetchRandomValue simply returned the output of _fetchPersonalisedSeed(_randomRequestId), there is a risk that a game studio would mistakenly:

- Call _fetchRandomValue
- The separately call _fetchRandomValues

Returning the pre-image of a set of random values. It would allow someone to simply generate the output of _fetchRandomValues.

I realise that given the platform is a blockchain, once the seed from the RandomSeedProvider is locked in, anyone can guess the values returned by _fetchRandomValue and _fetchRandomValues. However, hashing before outputting makes it harder.